Load packages.
```{r}
#| output: false
#| code-fold: true
#| code-summary: "Show the code"

library("tidymodels")
tidymodels::tidymodels_prefer()
```

# Applied Biostatistical Methods {#sec-stat_intro}

In this part, each of the following parts delve into an area of biostatistics. This includes a chapter on diversity measures, hypothesis testing and several chapters on models such as Ordinary Linear Regression. Each chapter is structured in a similar way. First, a commonly occurring problem is introduced, followed by an explanation of the method and how it solves the problem. Then, the math behind the method is then explained, and lastly, a visual representation. Deviations from this structure may occur when e.g. a text based introduction is insufficient.

In the first part of the book, the `iris` data set was used to exemplify the usage of the `tidymodels` package. Some of the following chapters reuses the `iris` data ses wheras some uses the also common `mtcars` data set. For the chapters pertaining to diversity measures, a count matrix from a metagenomics study is used instead. The raw data can be obtained from [@Cox2021]. Each row is a sample from a person, each column is a genus and the values represent the abundance of said genus in said sample. Explaining how the count matrix is acquired from raw sequences is out of scope for this book. Scripts used for the preprocessing exists in the [GitHub repository](https://github.com/WilliamH-R/BioStatistics) from which this book is created, located in the `src/` sub-folder.

A snippet of the data set can be seen below:

```{r}
count_matrix <- readr::read_rds("https://github.com/WilliamH-R/BioStatistics/raw/main/data/count_matrix/count_matrix.rds")
count_matrix
```

To speed up data processing, and to make results less overwhelming, a subset of the count matrix is used at times as test data when exemplifying.
```{r}
count_matrix_test <- count_matrix |> 
  slice_head(n = 10) |> 
  select(Sample, Actinomyces,
         Adlercreutzia, Agathobacter, Akkermansia)
count_matrix_test
```

Microbiome data is inherently compositional, meaning that the sum of all genera in a sample is constant, so the count numbers indicate the proportion of a specific genus. As a consequence of compositional data, commonly used statistical tools are not applicable. To avoid this compositionality, the count matrix is transformed using the centered log-ratio (clr) transformation.

```{r}
count_matrix_clr <- readr::read_rds("https://github.com/WilliamH-R/BioStatistics/raw/main/data/count_matrix/count_matrix_clr.rds")
count_matrix_clr
```

# Session Info

```{r}
sessioninfo::session_info()
```